processors/communicationdetails-deriver/src/main/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriver.java

```diff
@@ -184,32 +187,46 @@ public CommunicationDetails processOneToOne(String tenantId, Trace item) throws
-        for (int i = 0; i < consumerNodes.size(); i++) {
-            Node n = consumerNodes.get(i);
-            if (n.getClass() == Producer.class) {
-                CommunicationDetails.Outbound ob = new CommunicationDetails.Outbound();
-                for (int j = 0; j < ((Producer) n).getCorrelationIds().size(); j++) {
-                    CorrelationIdentifier ci = ((Producer) n).getCorrelationIds().get(j);
-                    if (ci.getScope() == Scope.Interaction) {
-                        ob.getIds().add(ci.getValue());
-                    }
-                }
-                // Only record if outbound ids found
-                if (!ob.getIds().isEmpty()) {
-                    // Check if pub/sub
-                    ob.setMultiConsumer(((Producer) n).multipleConsumers());
-
-                    ob.setProducerOffset(TimeUnit.MILLISECONDS.convert((n.getBaseTime() - baseTime),
-                            TimeUnit.NANOSECONDS));
-                    cd.getOutbound().add(ob);
+    protected static void initialiseOutbound(Node n, long baseTime, CommunicationDetails cd,
+            StringBuilder nodeId) {
+        CommunicationDetails.Outbound ob = new CommunicationDetails.Outbound();
+        ob.getIds().add(nodeId.toString());
+        ob.setMultiConsumer(true);
+        ob.setProducerOffset(TimeUnit.MILLISECONDS.convert((n.getBaseTime() - baseTime),
+                    TimeUnit.NANOSECONDS));
+        cd.getOutbound().add(ob);
+
+        if (n.getClass() == Producer.class) {
+            ob = new CommunicationDetails.Outbound();
+            for (int j = 0; j < ((Producer) n).getCorrelationIds().size(); j++) {
+                CorrelationIdentifier ci = ((Producer) n).getCorrelationIds().get(j);
+                if (ci.getScope() == Scope.Interaction || ci.getScope() == Scope.ControlFlow) {
+                    ob.getIds().add(ci.getValue());
                 }
-            } else if (n.containerNode()) {
-                initialiseOutbound(((ContainerNode) n).getNodes(), baseTime, cd);
+            }
+            // Only record if outbound ids found
+            if (!ob.getIds().isEmpty()) {
+                // Check if pub/sub
+                ob.setMultiConsumer(((Producer) n).multipleConsumers());
+
+                ob.setProducerOffset(TimeUnit.MILLISECONDS.convert((n.getBaseTime() - baseTime),
+                        TimeUnit.NANOSECONDS));
+                cd.getOutbound().add(ob);
+            }
+        } else if (n.containerNode()) {
+            for (int i = 0; i < ((ContainerNode)n).getNodes().size(); i++) {
+                int len = nodeId.length();
+                nodeId.append(':');
+                nodeId.append(i);
+                initialiseOutbound(((ContainerNode) n).getNodes().get(i), baseTime, cd, nodeId);
+
+                // Remove this child's specific path, so that next iteration will add a different path number
+                nodeId.delete(len, nodeId.length());
```

Comment:
Thought it might be more efficient, as effectively just manipulating an index rather than creating a new StringBuilder for each node in the hierarchy?